# This script will generate jobs to run aneufinder on multiple input folders (eg sequencing runs
args <- commandArgs(trailingOnly=TRUE)
options(scipen = 999)

# The samplesheet should contain the following columns: Run (eg the name for the folder containing the bams), Method (Strand-seq or WGS)
samplesheet_file <- args[1]
# Input folder should contain the Run_Aneufinder.R scripts. Jobs will be written to [input_folder]/Jobs/ and logs to [input_folder]/Logs/
input_folder <- args[2] 
# Aneufinder output data will be written to this file. A new subfolder for each run will be generated.
output_folder <- args[3]
# The bam_folder should contain subfolders for each run containing bam files. [bam_folder]/[run]/*.bam
bam_folder <- args[4]
# 
bin_size <- args[5]
# Blacklist can be generated by Generate_Aneufinder_Blacklist.R script
blacklist_path <- args[6]
# The runs aneufinder will analyze, leave empty if you want to run for all runs in the samplesheet
runs <- args[7]

## Settings
if(file.exists(blacklist_path) == FALSE){
  stop(paste("! Blacklist not found: ", blacklist_path, sep =""))
}
# use BSgenome.Btaurus.UCSC.bosTau8 for UMD3.1 or use hg19 for human
reference_genome <- "BSgenome.Btaurus.UCSC.bosTau8"
# Multiple methods can be run (use comma to separate)
aneufinder_methods <- "edivisive,HMM,dnacopy"
# Multiple bin sizes can be run (c(100000, 200000))
if(!is.null(bin_size) == TRUE){
  aneufinder_bin_size <- bin_size
} else {
  aneufinder_bin_size <- 1000000
}
# Jobs and logs will be written to these folders:
logs_folder <- paste(input_folder, "Logs/", sep = "")
jobs_folder <- paste(input_folder, "Jobs/", sep = "")
dir.create(logs_folder, showWarnings = F)
dir.create(jobs_folder, showWarnings = F)


if(file.exists(samplesheet_file) == FALSE){
  stop(paste("! Samplesheet not found: ", samplesheet_file, sep =""))
}
samplesheet <- read.delim(samplesheet_file, header = T, stringsAsFactors = F)
# Remove positive and negative controls:
samplesheet <- samplesheet[samplesheet$Embryo != "POS" & samplesheet$Embryo != "NEG",]

# only select the specified runs
if(!is.na(runs) == TRUE){
  print(runs)
  samplesheet <- samplesheet[samplesheet$Run %in% runs,]
}

dir.create(output_folder, showWarnings = F)

Run_Aneufinder_Script <- paste(input_folder, "Run_Aneufinder.R", sep = "")

if(file.exists(Run_Aneufinder_Script)){
  
  for(run in unique(samplesheet$Run)){
    
    run_info <- samplesheet[which(samplesheet$Run == run),]
    bam_folder_run <- paste(bam_folder, run, sep = "")
    method <- ifelse(unique(run_info$Method) == "Strand-seq", TRUE, FALSE)
    
    if( length(list.files(bam_folder_run, pattern = "\\.bam$")) > 0){
      for(bin_size in aneufinder_bin_size){
        job_ID <- paste(jobs_folder, "Run_Aneufinder_", run, "_",bin_size, ".sh", sep = "")
        shell_script <- paste("
#!/bin/bash

#$ -S /bin/bash
#$ -l h_vmem=30G
#$ -l h_rt=04:00:00
#$ -cwd
#$ -o ",logs_folder, format(Sys.Date(), "%Y%m%d"),"_Run_Aneufinder_", run, "_",bin_size,"_log.txt
#$ -e ",logs_folder, format(Sys.Date(), "%Y%m%d"),"_Run_Aneufinder_", run, "_",bin_size,"_log.txt

# Folder containing all the bam files:
BAM_FOLDER=",bam_folder_run,"

# Output will be written to this folder:
OUTPUT_FOLDER=",output_folder, run, "/

# Bin size for aneufinder. For multiple bin sizes use , as separation: 500000,1000000
BIN_SIZE=",format(bin_size, scientific = FALSE),"

# Set to TRUE if this is a strand-seq analysis:
STRANDSEQ=",method,"

# Reference genome. Use hg19 for human. Use BSgenome.Btaurus.UCSC.bosTau8 for bovine.
REFERENCE=",reference_genome,"

# Blacklist to exclude regions
BLACKLIST=",blacklist_path,"

# Methods used for copy number state calling. Options: edivisive,HMM,dnacopy
METHODS=",aneufinder_methods,"

# Leave empty to use all copy number state. Else use (split on ,): zero-inflation,0-somy,1-somy,2-somy,3-somy,4-somy
CNV_STATES=\"\"

# Leave empty if you want to run for all chromosomes (1-22+X for hg19, 1-29+X for BSgenome.Btaurus.UCSC.bosTau8). Else use , to split (for example 1,2,3,4,5)
CHROMOSOMES=\"\"

date

guixr load-profile ~/.guix-profile/ -- <<EOF

	Rscript ",Run_Aneufinder_Script," $BAM_FOLDER $OUTPUT_FOLDER $BIN_SIZE $STRANDSEQ $REFERENCE $BLACKLIST $METHODS $CNV_STATES $CHROMOSOMES

EOF
                              
date", sep = "")
    
    print(paste("# Writing shell script: ",job_ID, sep = ""))
    write.table(shell_script, file = job_ID, sep = "\t", quote = F, row.names = F, col.names = F)
    
    command <- paste("qsub ", job_ID, sep ="")
    print(command)
    system(command)
      }
      } else {
        print("! No BAM files found in folder: ",bam_folder_run, " !", sep = "")
      }
    }
  } else {
    print(paste("! Aneufinder script: ", Run_Aneufinder_Script, " not found!"), sep = "")
}

